sdk/scheduler/src/main/java/org/apache/mesos/offer/constrain/RoundRobinByAttributeRule.java

```diff
@@ -0,0 +1,197 @@
+package org.apache.mesos.offer.constrain;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+import org.apache.commons.lang3.builder.EqualsBuilder;
+import org.apache.commons.lang3.builder.HashCodeBuilder;
+import org.apache.mesos.Protos.Attribute;
+import org.apache.mesos.Protos.Offer;
+import org.apache.mesos.Protos.TaskInfo;
+import org.apache.mesos.offer.AttributeStringUtils;
+import org.apache.mesos.offer.OfferRequirement;
+import org.apache.mesos.offer.TaskUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+/**
+ * Implements logic for Marathon's GROUP_BY operator for attribute values. Ensures that tasks are
+ * evenly distributed across distinct attribute values in the system as they are rolled out.
+ *
+ * Example:
+ *  attribute |     tasks
+ * -----------+---------------
+ *    rack:1  | a-1, b-1, c-1
+ *    rack:2  | a-2, c-2, c-3
+ *    rack:3  | b-2, c-4
+ * Result:
+ *  allow rack:3 only, unless we know that there's >3 racks via the attribute_count parameter
+ *
+ * Example:
+ *  attribute |     tasks
+ * -----------+---------------
+ *    rack:1  | a-1, b-1, c-1
+ *    rack:2  | a-2, c-2, c-3
+ *    rack:3  | b-2, c-4, b-3
+ * Result:
+ *  allow any of rack:1/rack:2/rack:3, unless we know that there's >3 racks via the attribute_count
+ *  parameter.
+ */
+public class RoundRobinByAttributeRule implements PlacementRule {
```

Comment:
Strictly speaking these RoundRobin rules are not taking into account very much Cluster state (which we could know).  Correct me if I'm wrong, but as long as this Offer has the attribute in question and is below the maxCount for this attribute we're going to to say it's OK even if the next Offer, in the Offers we've received would be better.  e.g.  max is 3, and this one has 2 on it already, but the next Offer in the stream has none.  I wouldn't block on this, but it's worth filing a ticket for future improvement such that we look at a wider snapshot of the world and make smarter decisions.  Of course optimistic offers will alter this greatly, as we will have very close to a global view of the cluster...